src/ota_proxy/cache_streaming.py

```diff
@@ -177,42 +147,51 @@ async def _provider_write_cache(self) -> AsyncGenerator[int, bytes]:
         """
         logger.debug(f"start to cache for {self.meta=}...")
         try:
-            if not self.meta:
-                raise ValueError("called before provider tracker is ready, abort")
-
             async with aiofiles.open(self.fpath, "wb", executor=self._executor) as f:
-                # let writer become ready when file is open successfully
-                self._writer_ready.set()
-
                 _written = 0
                 while _data := (yield _written):
                     if not self._space_availability_event.is_set():
-                        logger.warning(
-                            f"abort writing cache for {self.meta=}: {StorageReachHardLimit.__name__}"
-                        )
-                        self._writer_failed.set()
-                        return
+                        _err_msg = f"abort writing cache for {self.meta=}: {StorageReachHardLimit.__name__}"
+                        logger.warning(_err_msg)
+                        raise StorageReachHardLimit(_err_msg)
 
                     _written = await f.write(_data)
+
+                    # set the writer is ready on first chunk of data written,
+                    # signal the subcriber that the cache stream starts.
+                    if not self._writer_ready.is_set():
+                        self._writer_ready.set()
+
                     self._bytes_written += _written
 
-            logger.debug(
-                "cache write finished, total bytes written"
-                f"({self._bytes_written}) for {self.meta=}"
-            )
+                # NOTE: cover the case if the requested file has 0 size:
+                self._writer_ready.set()
+
+            # logger.debug(
+            #     "cache write finished, total bytes written"
+            #     f"({self._bytes_written}) for {self.meta=}"
+            # )
+            # NOTE(20240805): mark the writer succeeded in advance to release the
+            #   subscriber faster. Whether the database entry is committed or not
+            #   doesn't matter here, the subscriber doesn't need to fail if caching
+            #   finished but db commit failed.
+            self._writer_finished.set()
```

Comment:
Refinement: release subscriber faster by marking the writer as finished right after caching finished without waiting for the db entry committed.